src/components/form-control/form-control.js

```diff
@@ -0,0 +1,97 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+// @flow
+import * as React from 'react';
+import {getOverride, getOverrideProps} from '../../helpers/overrides';
+import {STYLETRON_PROP_MAPPER} from './constants';
+import {
+  Label as StyledLabel,
+  Caption as StyledCaption,
+  ControlContainer as StyledControlContainer,
+} from './styled-components';
+import type {FormControlPropsT} from './types';
+
+function mapStyletronProps(props: {}, mapper: {}) {
+  return Object.keys(props).reduce((newProps, propName) => {
+    const newName = mapper[propName] && `$${propName}`;
+    if (newName) {
+      newProps[newName] = props[propName];
+    }
+    return newProps;
+  }, {});
+}
+
+export default class FormControl extends React.Component<FormControlPropsT> {
+  static defaultProps = {
+    overrides: {},
+    label: null,
+    caption: null,
+    error: false,
+  };
+
+  render() {
+    const {
+      overrides: {Label: LabelOverride, Caption: CaptionOverride},
+      label,
+      caption,
+      error,
+      children,
+    } = this.props;
+
+    const onlyChildProps = React.Children.only(children).props;
+    const sharedProps = mapStyletronProps(
+      onlyChildProps,
+      STYLETRON_PROP_MAPPER,
+    );
+    sharedProps.$error = this.props.error || sharedProps.$error;
+    const Label = getOverride(LabelOverride) || StyledLabel;
+    const Caption = getOverride(CaptionOverride) || StyledCaption;
+
+    return (
+      <React.Fragment>
+        {label && (
+          <Label
+            htmlFor={onlyChildProps.id}
+            {...sharedProps}
+            {...getOverrideProps(LabelOverride)}
+          >
+            {typeof label === 'function' ? label(sharedProps) : label}
+          </Label>
+        )}
+        <StyledControlContainer>
+          {children}
+          {(caption || error) && (
+            <Caption {...sharedProps} {...getOverrideProps(CaptionOverride)}>
+              {error && typeof error !== 'boolean'
+                ? typeof error === 'function' ? error(sharedProps) : error
+                : typeof caption === 'function'
+                  ? caption(sharedProps)
+                  : caption}
+            </Caption>
+          )}
+        </StyledControlContainer>
```

Comment:
It looks like StyledControlContainer has a margin-bottom by default. I think this makes sense, but we definitely want to make it easy to remove this margin if necessary, and I think the best way to do that is expose the Container via overrides.

Also curious, why leave label outside the container, but put caption inside? Maybe it makes sense to put both inside and just make the Container the root element (instead of using fragment)

Reply:
Oh, it definitely has to be a part of overrides! I added it last minute and totally missed to add to the overrides. The label is left outside in order to make it easier to move label around, for example, place it on the left while a caption always aligned with it's control.